### DISCUSSION TITLE
Can order of D0-D7 pins in 8-bit parallel mode be reversed?

### DISCUSSION THREAD

**Discussion by sandric** [Bodmer/TFT_eSPI#2458] [General]
I have a small issue with my setup - I'm using RP2040 MCU, and would like to define my D0-D7 pins not in ascending order but in descending. This is due to my pcb layout - without it cannot be made one-sided. So I wonder is there any way I can define PIO pins directly, if its PIO related issue, I'm not sure? Or change somehow that it will go in reverse? I can change it in library code, and use those changes locally only if its smth that is hard to code or make default etc.

Here's my `UserSetup.h` :

```C
#define USER_SETUP_ID 303
#define TFT_PARALLEL_8_BIT
#define SSD1963_800ALT_DRIVER
#define TFT_RGB_ORDER TFT_BGR

#define TFT_DC   7
#define TFT_WR   6
#define TFT_RST  5

// This works flawlessly:
#define TFT_D0   8
#define TFT_D1   9
#define TFT_D2   10
#define TFT_D3   11
#define TFT_D4   12
#define TFT_D5   13
#define TFT_D6   14
#define TFT_D7   15

// This doesn't work, but its actually how I need to define pins:
/*
#define TFT_D0   15
#define TFT_D1   14
#define TFT_D2   13
#define TFT_D3   12
#define TFT_D4   11
#define TFT_D5   10
#define TFT_D6   9
#define TFT_D7   8
*/

#define LOAD_GLCD
#define LOAD_FONT2
#define LOAD_FONT4
#define LOAD_FONT6
#define LOAD_FONT7
#define LOAD_FONT8
#define LOAD_GFXFF
#define SMOOTH_FONT
```

**Comment by Bodmer** [comment]
Reversing the order is not supported by the library. The code changes to make it possible are not trivial. Wire links would be simpler to add.

**Reply by sandric** [reply]
Ok, fair enough, thx.

**Comment by troisiemetype** [comment]
RP2040 PIO registers has the ability to shift data out LSB first, or MSB first (See chapter _3.2.3 - Registers_ of RP2040 datasheet). Once set, you would have to re-write the PIO program so that instead of discarding the 24 MSB you don't care about, then the remaining 8 LSB, you would first output the 8 LSB, then discard the remaining.

**Reply by Bodmer** [reply]
Also all 16 bit values sent will need to be byte swapped.
For the 32 bit writes used by setWindow the order of the 4 bytes must also be reversed.

**Reply by troisiemetype** [reply]
Yes, I understood that after commenting. I've spent part of yesterday trying to understand your pio code. I think I almost got it. :)

**Reply by sandric** [reply]
As pointed out by @Bodmer - its easier to change design, and I've already designed the way it will work without need to swap pins directions. If you interested in this for future use by anyone else - sure please look into it, just wanted to point out I already redesigned my circuit :smiley:
